### ISSUE TITLE
`no-useless-undefined` makes code unreadable and conflicts with types (required parameters)

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#1533]
Related:

- Possibly identical to: #1377 
- Also related to: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1378

But I wanted to open a more generic issue to argue against the rule as a whole… probably.

In short, just because `undefined` is the same as "the lack of value", the code resulting from this rule might not be as readable.

```js
t.equals(thisShouldBeUndefined, undefined);
                                ^ Do not use useless `undefined`. eslint(unicorn/no-useless-undefined)
```

Now I _fix_ this to:

```js
t.equals(thisShouldBeUndefined);
                              ^ Expected 2-4 arguments, but got 1
                                index.d.ts(159, 29): An argument for 'expected' was not provided.
```

even without the types, the function does not read _right:_ `X must equal to… what?`

- Is `t.equals` an exception or is it the rule?
- Should `no-useless-undefined` exclude required TypeScript parameters?
- Can the rule work in plain JavaScript where the above isn't detectable?

**Comment by sindresorhus** [comment]
> Is t.equals an exception or is it the rule?

It's an exception.

> Should no-useless-undefined exclude required TypeScript parameters?

No. This rule is not very useful for TS projects anyway, so I would just disable it there. I guess we could disable it in https://github.com/xojs/eslint-config-xo-typescript

> Can the rule work in plain JavaScript where the above isn't detectable?

In most cases, `undefined` and lack of value are interchangeable. It's really only with unit testing it's not true.

**Comment by sindresorhus** [comment]
> But I wanted to open a more generic issue to argue against the rule as a whole… probably.

Do note that the rule catches other cases than function calls, which seems to what you're objecting to.

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-useless-undefined.md

**Comment by fregante** [comment]
I just noticed `checkArguments`, which appears to work for me (and thus would probably prefer it to default to `false` 😜)
